WinImageBuilder.psm1

```diff
@@ -409,20 +531,213 @@ function Add-VirtIODriversFromISO($vhdDriveLetter, $image, $isoPath) {
-                     ([System.IO.Path]::GetFileNameWithoutExtension($path))
+function Get-PathWithoutExtension {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$Path
+    )
+    return Join-Path ([System.IO.Path]::GetDirectoryName($Path)) `
+                     ([System.IO.Path]::GetFileNameWithoutExtension($Path))
 }
 
-function New-WindowsCloudImage()
-{
+function Compress-Image {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$VirtualDiskPath,
+        [Parameter(Mandatory=$true)]
+        [string]$ImagePath
+    )
+    if (!(Test-Path $VirtualDiskPath)) {
+        Throw "$VirtualDiskPath not found"
+    }
+    $tmpName = $ImagePath + "." + (Get-Random)
+
+    $7zip = Join-Path $localResourcesDir 7za.exe
+    $pigz = Join-Path $localResourcesDir pigz.exe
+    try {
+        Write-Host "Archiving $VirtualDiskPath to tarfile $tmpName"
+        pushd ([System.IO.Path]::GetDirectoryName((Resolve-path $VirtualDiskPath).Path))
+        try {
+            # Avoid storing the full path in the archive
+            $imageFileName = (Get-Item $VirtualDiskPath).Name
+            Write-Host "Creating tar archive..."
+            & $7zip a -ttar $tmpName $imageFileName
+            if ($LASTEXITCODE) {
+                if ((Test-Path $imageFileName)) {
+                    Remove-Item -Force $imageFileName
+                }
+                throw "7za.exe failed while creating tar file for image: $tmpName"
+            }
+        } finally {
+            popd
+        }
+
+        Remove-Item -Force $VirtualDiskPath
+        Write-Host "Compressing $tmpName to gzip"
+        pushd ([System.IO.Path]::GetDirectoryName((Resolve-path $tmpName).Path))
+        try {
+            $tmpPathName = (Get-Item $tmpName).Name
+            Write-Host "Creating gzip..."
+            & $pigz -p12 $tmpPathName
+            if ($LASTEXITCODE) {
+                $gzipped = ($tmpPathName + ".gz")
+                if ((Test-Path $gzipped)) {
+                    Remove-Item -Force $gzipped
+                }
+                throw "pigz.exe failed while creating gzip file for : $tmpName"
+            }
+        } finally {
+            popd
+        }
+    } catch {
+        Remove-Item -Force $tmpName -ErrorAction SilentlyContinue
+        Remove-Item -Force $VirtualDiskPath -ErrorAction SilentlyContinue
+        throw
+    }
+    Move-Item ($tmpName + ".gz") $ImagePath
+    Write-Output "MaaS image is ready and available at: $ImagePath"
+}
+
+function Shrink-VHDImage {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [string]$VirtualDiskPath
+    )
+    Write-Host "Shrinking VHD to minimum size"
+
+    $vhdSize = (Get-VHD -Path $VirtualDiskPath).Size
+    $vhdSizeGB = $vhdSize/1GB
+    Write-Host "Initial VHD size is: $vhdSizeGB GB"
+
+    $Drive = (Mount-VHD -Path $VirtualDiskPath -Passthru | Get-Disk | Get-Partition | Get-Volume).DriveLetter
+    Optimize-Volume -DriveLetter $Drive -Defrag -ReTrim

```

Comment:
As presented here, retrim is okay: https://technet.microsoft.com/en-us/library/hh848675(v=wps.630).aspx